File: epi_judge_cpp/apply_permutation.cc
#include <vector>

using std::vector;

void ApplyPermutation(vector<int>* perm_ptr, vector<int>* A_ptr) {
  // Implement this placeholder.
  return;
}

vector<int> ApplyPermutationWrapper(vector<int> perm, vector<int> A) {
  ApplyPermutation(&perm, &A);
  return A;
}

#include "test_framework/generic_test.h"

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"perm", "A"};
  GenericTestMain(args, "apply_permutation.tsv", &ApplyPermutationWrapper,
                  DefaultComparator{}, param_names);
  return 0;
}

File: epi_judge_python_solutions/apply_permutation.py
from test_framework import generic_test


def apply_permutation(perm, A):

    for i in range(len(A)):
        # Check if the element at index i has not been moved by checking if
        # perm[i] is nonnegative.
        next = i
        while perm[next] >= 0:
            A[i], A[perm[next]] = A[perm[next]], A[i]
            temp = perm[next]
            # Subtracts len(perm) from an entry in perm to make it negative,
            # which indicates the corresponding move has been performed.
            perm[next] -= len(perm)
            next = temp
    # Restore perm.
    perm[:] = [a + len(perm) for a in perm]


def apply_permutation_wrapper(perm, A):
    apply_permutation(perm, A)
    return A


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("apply_permutation.py",
                                       "apply_permutation.tsv",
                                       apply_permutation_wrapper))
